### PR TITLE
Ensure a base space always exists (+ fix gRPC sub-environments)

### LINKED ISSUE #3221: Sub-environment variables don't work with gRPC
**Describe the bug**
I recently discovered that the latest version of Insomnia supports environment variables; however I am unable to get this to work. When I put in an environment variable of `127.0.0.1:6565` and attempt to use this, nothing happens in the UI (there are errors in the dev console). If I directly paste `127.0.0.1:6565` in, it works as expected (old behavior). I have tried downgrading to 2020.1.0 but that doesn't fix the issue either. 

**To Reproduce**
In the environment tab, add a new variable. Fill that variable in and make a gRPC request.

**Expected behavior**
The request should be successful, but it is not

**Screenshots**
(These are intentionally cropped)
image
image
image

image

**Desktop (please complete the following information):**
Version: Insomnia 2021.1.1
Release date: 3/9/2021
OS: Darwin x64 19.6.0
Electron: 9.1.1
Node: 12.14.1
V8: 8.3.110.13-electron.0
Architecture: x64
node-libcurl: libcurl/7.69.1 OpenSSL/1.1.1g zlib/1.2.11 brotli/1.0.7 libidn2/2.1.1 libssh2/1.9.0 nghttp2/1.41.0

**Additional context**
Error message:
```
18:24:20.309 › Unhandled Promise PromiseRejectionEvent {isTrusted: true, promise: Promise, reason: Error: attempted to output null or undefined value
    at file:///Applications/Insomnia.app/Content…, type: "unhandledrejection", target: global, …}bubbles: falsecancelBubble: falsecancelable: truecomposed: falsecurrentTarget: global {parent: global, opener: null, top: global, length: 0, frames: global, …}defaultPrevented: falseeventPhase: 0isTrusted: truepath: [global]promise: Promise__proto__: Promise[[PromiseStatus]]: "rejected"[[PromiseValue]]: Error: attempted to output null or undefined value
    at file:///Applications/Insomnia.app/Contents/Resources/app.asar/bundle.js:37548:26
    at /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/nunjucks/src/environment.js:41:5
    at RawTask.call (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/asap.js:40:19)
    at flush (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/raw.js:50:29)
    at processTicksAndRejections (internal/process/task_queues.js:76:11)location: {line: 1, column: 1}path: "url"reason: "undefined"type: "render"message: "attempted to output null or undefined value"stack: "Error: attempted to output null or undefined value↵    at file:///Applications/Insomnia.app/Contents/Resources/app.asar/bundle.js:37548:26↵    at /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/nunjucks/src/environment.js:41:5↵    at RawTask.call (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/asap.js:40:19)↵    at flush (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/raw.js:50:29)↵    at processTicksAndRejections (internal/process/task_queues.js:76:11)"__proto__: Errorreason: Error: attempted to output null or undefined value
    at file:///Applications/Insomnia.app/Contents/Resources/app.asar/bundle.js:37548:26
    at /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/nunjucks/src/environment.js:41:5
    at RawTask.call (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/asap.js:40:19)
    at flush (/Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/asap/raw.js:50:29)
    at processTicksAndRejections (internal/process/task_queues.js:76:11)returnValue: truesrcElement: global {parent: global, opener: null, top: global, length: 0, frames: global, …}target: global {parent: global, opener: null, top: global, length: 0, frames: global, …}timeStamp: 595062.414999993type: "unhandledrejection"__proto__: PromiseRejectionEvent
setTimeout (async)
consoleLog @ /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/electron-log/src/transports/console.js:93
transport @ /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/electron-log/src/transports/console.js:46
runTransport @ /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/electron-log/src/log.js:44
runTransports @ /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/electron-log/src/log.js:27
log @ /Applications/Insomnia.app/Contents/Resources/app.asar/node_modules/electron-log/src/log.js:21
```

### PR DESCRIPTION
**Pull request by develohpanda**
### Changes
* The export functions now always require a list of workspaces (the magic where we default an array of `[null]` has been removed
	* This directly impacts the plugin API but a fallback has been put into place, which we might need to add metrics for (see self-review)
* Converted some of the render function arguments into an object to better cater for optionals
	* As a positive side-effect, fixing these types identified an error in the types [ref](https://github.com/Kong/insomnia/pull/3819#discussion_r674502952). Fixing the types fixes #3221
* If a workspace did not have a `parentId` it is set to the `BASE_SPACE_ID`
* Some existing tests need to call `models.space.all()` to make sure the parent space exists; this is done in `beforeEach` blocks

Please see self-review comments for context in different areas

### Demos
* [Don't export if there are no collections in the selected space ](https://www.loom.com/share/0f76f8b7ee1c4964880b8afbd3e075f8)

Closes INS-812

### CODE REVIEW

File: packages/insomnia-app/app/network/grpc/prepare.ts [hunk 1]
```diff
@@ -23,15 +23,16 @@ export const prepareGrpcRequest = async (
...
-    // @ts-expect-error -- TSCONVERSION req can be null but should not try to render if it is null
-    req,
-    environment,
-    RENDER_PURPOSE_SEND,
-    {},
-    canClientStream(methodType),
+    {
+      // @ts-expect-error -- TSCONVERSION req can be null but should not try to render if it is null
+      request: req,
+      environmentId,
```
Comment (develohpanda) [inline packages/insomnia-app/app/network/grpc/prepare.ts:30]: Previously we send the entire environment, but it only expects an environmentId, which fixes https://github.com/Kong/insomnia/issues/3221. Yay for improved types!

File: packages/insomnia-app/app/plugins/context/data.ts [hunk 2]
```diff
@@ -23,7 +25,10 @@ const buildImportRawConfig = (options: PluginImportOptions): ImportRawConfig =>
   ),
 });
 
-export const init = () => ({
+// TODO: add metrics here to track how frequently this fallback is being used
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/data.ts:29]: How can we add a metric here to check this - through segment?

File: packages/insomnia-app/app/plugins/context/data.ts [hunk 3]
```diff
@@ -39,7 +43,7 @@ export const init = () => ({
         includePrivate,
         format,
       }: InsomniaExport = {}) => exportWorkspacesData(
-        workspace ? [workspace] : [],
+        workspace ? [workspace] : await getWorkspacesInActiveSpace(activeSpaceId),
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/data.ts:46]: Initially, sending no workspace would cause all workspaces to export

After introducing spaces, sending no workspace would cause all workspaces _in the base space only, always_ to export (this is because the export logic would fetch all of the workspaces with a `null` parentId - thus those in the base space).

After spaces, the behavior wasn't really ideal - what should happen is if the plugin does not specify which workspace to export, we should by default get all of the workspaces in the active space. This PR updates that behavior by sending the `activeSpaceId` whenever this plugin context is initialized, and a large majority of the changes in this PR are related to getting the space to this `init` function

File: packages/insomnia-app/app/network/network.ts [hunk 4]
```diff
@@ -1003,7 +1005,7 @@ async function _applyRequestPluginHooks(
   for (const { plugin, hook } of await plugins.getRequestHooks()) {
     const context = {
       ...(pluginContexts.app.init(RENDER_PURPOSE_NO_RENDER) as Record<string, any>),
-      ...pluginContexts.data.init(),
+      ...pluginContexts.data.init(renderedContext.getSpaceId()),
```
Comment (develohpanda) [inline packages/insomnia-app/app/network/network.ts:1009]: This follows the same pattern as fetching the environment Id four lines down when initializing the network context

File: packages/insomnia-app/app/plugins/context/data.ts [hunk 5]
```diff
@@ -39,7 +47,7 @@ export const init = () => ({
         includePrivate,
         format,
       }: InsomniaExport = {}) => exportWorkspacesData(
-        workspace ? [workspace] : [],
+        workspace ? [workspace] : await getWorkspaces(activeSpaceId),
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/data.ts:52]: Initially, sending no workspace would cause all workspaces to export

After introducing spaces, sending no workspace would cause all workspaces _in the base space only, always_ to export (this is because the export logic would fetch all of the workspaces with a `null` parentId - thus those in the base space).

The behavior isn't great - what should happen is if the plugin does not specify which workspace to export, we should by default get all of the workspaces in the _active_ space, which matches more closely to existing behavior.

This PR updates that behavior by sending the `activeSpaceId` whenever this plugin context is initialized, and a large majority of the changes in this PR are related to getting the space to this `init` function.

File: packages/insomnia-app/app/plugins/context/data.ts [hunk 6]
```diff
@@ -23,7 +24,14 @@ const buildImportRawConfig = (options: PluginImportOptions): ImportRawConfig =>
...
 
-export const init = () => ({
+// TODO: add metrics here to track how frequently this fallback is being used
+const getWorkspaces = (activeSpaceId?: string) => 
+  activeSpaceId 
+    ? models.workspace.findByParentId(activeSpaceId) 
+    : models.workspace.all();
+
+// Only in the case of running unit tests from Inso via send-request can activeSpaceId be undefined. This is because the concept of a space doesn't exist in git/insomnia sync or an export file
+export const init = (activeSpaceId?: string) => ({
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/data.ts:36]: The only time an `activeSpaceId` is not available is when we are coming from inso (which has no context of a space) through `insomnia-send-request`, and in that case, we return all of the workspaces present. This is only for completeness; it is unlikely there is a workflow that routes through here via inso but in case there is we have an implementation.

File: packages/insomnia-app/app/common/__tests__/har.test.ts [hunk 7]
```diff
@@ -6,543 +6,544 @@ import { AUTH_BASIC } from '../constants';
 import * as harUtils from '../har';
 import * as render from '../render';
 
-describe('exportHar()', () => {
-  beforeEach(globalBeforeEach);
+describe('export', () => {
```
Comment (develohpanda) [inline packages/insomnia-app/app/common/__tests__/har.test.ts:9]: Some of the unit testing diffs are not lining up well in GitHub - if that is the case, it is most likely because of a top-level describe block that has been introduced (with a common `beforeEach`) causing an indent change on each line. Be sure to hide whitespace changes when reviewing: 
image

File: packages/insomnia-app/app/models/space.ts [hunk 8]
```diff
@@ -64,6 +63,13 @@ export function update(space: Space, patch: Partial<Space>) {
...
+export async function all() {
+  const spaces = await db.all<Space>(type);
+
+  if (!spaces.find(c => c._id === BASE_SPACE_ID)) {
+    await create({ _id: BASE_SPACE_ID, name: getAppName(), remoteId: null });
+    return db.all<Space>(type);
+  }
+
+  return spaces;
 }
```
Comment (develohpanda) [inline packages/insomnia-app/app/models/space.ts:75]: The same pattern that existed previously for ensuring that a workspace always existed

File: packages/insomnia-app/app/common/__tests__/render.test.ts [hunk 9]
```diff
@@ -1,846 +1,745 @@
...
+import { Environment } from '../../models/environment';
+import { Workspace } from '../../models/workspace';
 import * as renderUtils from '../render';
 
 jest.mock('electron');
 
-describe('render()', () => {
-  beforeEach(globalBeforeEach);
+const envBuilder = createBuilder(environmentModelSchema);
+const reqGroupBuilder = createBuilder(requestGroupModelSchema);
```
Comment (develohpanda) [inline packages/insomnia-app/app/common/__tests__/render.test.ts:13]: Using these builders to fix many of the type errors in this file. This file hasn't been updated since flow and my current work was in this area.

File: packages/insomnia-app/app/common/database.ts [hunk 10]
```diff
@@ -549,7 +549,7 @@ export const database = {
 
     let docsToReturn: T[] = doc ? [doc] : [];
 
-    async function next(docs: T[]) {
+    async function next(docs: T[]): Promise<T[]> {
```
Comment (develohpanda) [inline packages/insomnia-app/app/common/database.ts:552]: TS didn't auto-detect this return type (and thus didn't type the parent type properly)

File: packages/insomnia-app/app/common/render.ts [hunk 11]
```diff
@@ -54,10 +54,17 @@ export type HandleGetRenderContext = () => Promise<RenderContextAndKeys>;
 export type HandleRender = <T>(object: T, contextCacheKey?: string | null) => Promise<T>;
 
 export async function buildRenderContext(
```
Comment (develohpanda) [inline packages/insomnia-app/app/common/render.ts:56]: In a future PR, we should really type the return type of this correctly; it will save us many headaches in the future

File: packages/insomnia-app/app/common/render.ts [hunk 12]
```diff
@@ -280,17 +287,31 @@ export async function render<T>(
...
-  purpose: RenderPurpose | null = null,
-  extraInfo: ExtraRenderInfo | null = null,
+  {
+    request,
+    environmentId,
+    ancestors: _ancestors,
+    purpose,
+    extraInfo,
+  }: RenderContextOptions,
 ): Promise<Record<string, any>> {
```
Comment (develohpanda) [inline packages/insomnia-app/app/common/render.ts:311]: In a future PR, we should really type the return type of this correctly; it will save us many headaches

File: packages/insomnia-app/app/plugins/context/__tests__/data.test.ts [hunk 13]
```diff
@@ -47,7 +52,7 @@ describe('app.import.*', () => {
         modified: 999,
         description: '',
         name: 'New',
-        parentId: null,
+        parentId: BASE_SPACE_ID,
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/__tests__/data.test.ts]: Should import into the current space

File: packages/insomnia-app/app/plugins/context/__tests__/data.test.ts [hunk 14]
```diff
@@ -95,7 +100,7 @@ describe('app.import.*', () => {
...
```
Comment (develohpanda) [inline packages/insomnia-app/app/plugins/context/__tests__/data.test.ts]: Should import into the current space

File: packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts [hunk 15]
```diff
@@ -66,11 +68,11 @@ describe('NeDBClient', () => {
       const wrk1Yml = path.join(GIT_INSOMNIA_DIR, models.workspace.type, 'wrk_1.yml');
       const req1Yml = path.join(GIT_INSOMNIA_DIR, models.request.type, 'req_1.yml');
       const reqXYml = path.join(GIT_INSOMNIA_DIR, models.request.type, 'req_x.yml');
-      const pNeDB = new NeDBClient('wrk_1');
+      const pNeDB = new NeDBClient('wrk_1', BASE_SPACE_ID);
       expect(YAML.parse((await pNeDB.readFile(wrk1Yml, 'utf8')).toString())).toEqual(
         expect.objectContaining({
           _id: 'wrk_1',
-          parentId: null,
+          parentId: null, // should be reset to default value instead of the active space id
```
Comment (develohpanda) [inline packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts:75]: This is because we set/clear the `parentId` when writing/reading through the git integration

File: packages/insomnia-app/app/ui/components/modals/space-settings-modal.tsx [hunk 16]
```diff
@@ -15,15 +16,9 @@ import ModalBody from '../base/modal-body';
...
-  space?: Space;
-}
-
-interface ReduxDispatchProps {
-  handleRemoveSpace: (space: Space) => void;
-}
+export type ReduxProps = ReturnType<typeof mapStateToProps> & ReturnType<typeof mapDispatchToProps>;
 
-interface Props extends ReduxStateProps, ReduxDispatchProps { }
+type Props = ReduxProps
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/components/modals/space-settings-modal.tsx:21]: Seems trivial, but just following our own set convention :)

File: packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts [hunk 17]
```diff
@@ -28,19 +29,19 @@ describe('selectors', () => {
...
       await models.space.create();
 
-      // set first as selected
-      const state = await reduxStateForTest({ activeSpaceId: null });
+      // set base as selected
+      const state = await reduxStateForTest({ activeSpaceId: undefined });
 
       const space = selectActiveSpace(state);
-      expect(space).toBeUndefined();
+      expect(space).toStrictEqual(expect.objectContaining<Partial<Space>>({ _id: BASE_SPACE_ID }));
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts:44]: Tests to make sure the base space is returned in cases where we can't find the actual expected space

File: packages/insomnia-app/app/ui/redux/modules/global.tsx [hunk 18]
```diff
@@ -461,10 +461,22 @@ const writeExportedFileToFileSystem = (filename: string, jsonData: string, onDon
...
+  const workspaces = selectWorkspacesForActiveSpace(state);
+
+  if (!workspaces.length) {
+    dispatch(loadStop());
+    showAlert({
+      title: 'Cannot export',
+      message: <>There are no workspaces to export in the <strong>{activeSpaceName}</strong> space.</>,
+    });
+    return;
+  }
```
Comment (develohpanda) [inline packages/insomnia-app/app/ui/redux/modules/global.tsx:475]: Demo linked in description

File: packages/insomnia-app/app/network/grpc/prepare.ts (hunk 1 above)
Comment (develohpanda) [reply packages/insomnia-app/app/network/grpc/prepare.ts:30]: image

File: packages/insomnia-app/app/common/database.ts (hunk 10 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/common/database.ts:552]: huh.  I fiddled earlier this morning but didn't see why either.

File: packages/insomnia-app/app/common/render.ts [hunk 19]
```diff
@@ -54,10 +54,17 @@ export type HandleGetRenderContext = () => Promise<RenderContextAndKeys>;
...
-  ancestors: BaseModel[] | null,
-  rootEnvironment: Environment | null,
-  subEnvironment: Environment | null,
-  baseContext: Record<string, any> = {},
+  {
+    ancestors,
+    rootEnvironment,
+    subEnvironment,
+    baseContext = {},
+  }: {
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/common/render.ts:62]: I'm always happy to see us moving to this option-argument way

File: packages/insomnia-app/app/common/render.ts [hunk 20]
```diff
@@ -373,24 +394,26 @@ export async function getRenderContext(
 
   baseContext.getEnvironmentId = () => environmentId;
 
+  // It is possible for a space to not exist because this
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/common/render.ts]: can you maybe reword this comment or clarify a bit more?

File: packages/insomnia-app/app/models/space.ts [hunk 21]
```diff
@@ -8,15 +9,13 @@ export const prefix = 'sp';
...
-export const isBaseSpace = ({ _id }: Pick<Space, '_id'>) => _id === BASE_SPACE_ID;
-export const isNotBaseSpace = (space: Pick<Space, '_id'>) => !isBaseSpace(space);
-export const isLocalSpace = ({ remoteId }: Pick<Space, 'remoteId'>) => remoteId === null;
-export const isRemoteSpace = (space: Pick<Space, 'remoteId'>) => !isLocalSpace(space);
-export const spaceHasSettings = (space: Pick<Space, 'remoteId' | '_id'>) => isLocalSpace(space) && !isBaseSpace(space);
+export const isBaseSpace = ({ _id }: Space) => _id === BASE_SPACE_ID;
+export const isNotBaseSpace = (space: Space) => !isBaseSpace(space);
+export const isLocalSpace = ({ remoteId }: Space) => remoteId === null;
+export const isRemoteSpace = (space: Space) => !isLocalSpace(space);
+export const spaceHasSettings = (space: Space) => isLocalSpace(space) && !isBaseSpace(space);
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/models/space.ts:18]: was it intentional to remove these `Pick`s?  They were there before because we don't always actually have a full space to pass to these functions.  E.g. in space-dropdown we have a partial space that we work with.

File: packages/insomnia-app/app/network/grpc/__tests__/prepare.test.ts [hunk 22]
```diff
@@ -10,6 +12,9 @@ import { prepareGrpcMessage, prepareGrpcRequest } from '../prepare';
 
 jest.mock('../../../common/render');
 
+const getRenderedGrpcRequest = mocked(_getRenderedGrpcRequest);
+const getRenderedGrpcRequestMessage = mocked(_getRenderedGrpcRequestMessage);
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/network/grpc/__tests__/prepare.test.ts:16]: noice.

File: packages/insomnia-app/app/network/grpc/__tests__/prepare.test.ts [hunk 23]
```diff
@@ -26,11 +31,12 @@ describe('prepareGrpcRequest', () => {
...
-        env,
-        RENDER_PURPOSE_SEND,
-        {},
-        false,
+        {
+          request: gr,
+          environmentId: env._id,
+          purpose: RENDER_PURPOSE_SEND,
+          skipBody: false,
+        },
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/network/grpc/__tests__/prepare.test.ts:39]: I know I already commented on it, but the readability improvement here by using an object argument is stunning.  great improvement.

File: packages/insomnia-app/app/network/grpc/prepare.ts (hunk 1 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/network/grpc/prepare.ts:30]: wow.  just wow.

File: packages/insomnia-app/app/plugins/context/data.ts (hunk 2 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/plugins/context/data.ts:29]: see my comment above regarding migrations, but I think migrations to a new insomnia format may be the best way.  we can put the analytics call right there in that migration function (from v4 -> v5)

File: packages/insomnia-app/app/plugins/context/data.ts (hunk 5 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/plugins/context/data.ts:52]: I think I actually follow all of that.  OK.  cool.

File: packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts (hunk 15 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts:75]: So a thought on that... what would it matter if we left the parentId there?  It wasn't used for anything before, right?

File: packages/insomnia-app/app/ui/components/dropdowns/space-dropdown.tsx [hunk 24]
```diff
@@ -109,25 +99,20 @@ export const SpaceDropdown: FC<Props> = ({ vcs }) => {
...
+
   return (
     <Dropdown renderButton={button} onOpen={refresh}>
-      <SpaceDropdownItem
-        isActive={baseSpace._id === activeSpace._id}
-        setActive={setActive}
-        space={baseSpace}
-      />
-
+      {spaces.filter(isBaseSpace).map(renderSpace)}
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/ui/components/dropdowns/space-dropdown.tsx:113]: I know we gotta get the show on the road, but though I'd mention ramda's partition for this use case.  most definitely do not need to do here and now just thought it was a good example of where it's useful.

```ts
const [[baseSpace], ...otherSpaces] = partition(isBaseSpace, spaces)
```

File: packages/insomnia-app/app/ui/components/modals/index.ts [hunk 25]
```diff
@@ -24,7 +23,7 @@ export function showPrompt(config: PromptModalOptions) {
   return showModal(PromptModal, config);
 }
 
-export function showAlert(config) {
+export function showAlert(config: AlertModalOptions) {
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/ui/components/modals/index.ts:26]: nice one.

File: packages/insomnia-app/app/ui/components/modals/space-settings-modal.tsx (hunk 16 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/ui/components/modals/space-settings-modal.tsx:21]: oh, I'm ALLL for it, don't worry.

File: packages/insomnia-app/app/ui/components/sidebar/sidebar-children.tsx [hunk 26]
```diff
@@ -117,6 +120,7 @@ class SidebarChildren extends PureComponent<Props> {
             request={child.doc}
             hotKeyRegistry={hotKeyRegistry} // Necessary for plugin actions on requests
             activeEnvironment={activeEnvironment}
+            activeSpace={activeSpace}
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/ui/components/sidebar/sidebar-children.tsx:123]: I'm assuming there's a deep reason that I'm missing the explains why we can't get this from redux?

From what I can tell: all of these have it:

Screenshot_20210723_150322

If there's just one thing I'd like to happen in this PR it'd be to get this from redux in all of these cases.  I looked at them and none of them appear to be connected already, which is a bummer, but we gotta start somewhere I suppose.

File: packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts (hunk 17 above)
Comment (dimitropoulos) [reply packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts:44]: with the context of where we were, where we are, and where we're going, this makes sense.  on the other hand if I step back and think about this it seems like this should throw an error in some way.  what is the (valid) situation in which there's no active space?

File: packages/insomnia-app/app/ui/redux/modules/workspace.ts [hunk 27]
```diff
@@ -61,8 +60,7 @@ export const createWorkspace = ({ scope, onCreate }: {
             {
               name,
               scope,
-              // @ts-expect-error TSCONVERSION the common parentId isn't typed correctly
```
Comment (dimitropoulos) [inline packages/insomnia-app/app/ui/redux/modules/workspace.ts:64]: man.. this PR is a trip.  when you add it all up there were QUITE a few places where we were dancing around this ghost space thingy.  glad we're addressing it sooner rather than later!

**Review by dimitropoulos — APPROVED** [review]
This PR is going to be a real pain in the tuchus to rebase, so I don't want to wait on it.

However, I'd like to maybe start thinking about creating a new version of the insomnia format that accounts for spaces.  That way we can ensure there will never be dangling workspaces (i.e. with null parentId like before).  And that way we only need to check once upon booting the app (and then run a small migration if so) and then never again.

If you do only one thing before merging: it'd be great to connect some of those components to redux so we can grab the active space (rather than prop drilling).  Apologies if I missed a reason why it's currently happening.

File: packages/insomnia-app/app/models/space.ts (hunk 21 above)
Comment (develohpanda) [reply packages/insomnia-app/app/models/space.ts:18]: Not anymore, the concept of a partial space is gone 😉

File: packages/insomnia-app/app/ui/components/sidebar/sidebar-children.tsx (hunk 26 above)
Comment (develohpanda) [reply packages/insomnia-app/app/ui/components/sidebar/sidebar-children.tsx:123]: You are correct these just come from redux, but I was hesitant to do that here because the PR was already quite far reaching and I wanted to reduce the changeset by following the simplest approach.

I'll attach the leaf components to redux in an immediate follow-up PR

File: packages/insomnia-app/app/ui/components/dropdowns/space-dropdown.tsx (hunk 24 above)
Comment (develohpanda) [reply packages/insomnia-app/app/ui/components/dropdowns/space-dropdown.tsx:113]: That'd be very useful! 🙌🏽 The more I come across use cases for ramda the more I want to use it

File: packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts (hunk 17 above)
Comment (develohpanda) [reply packages/insomnia-app/app/ui/redux/__tests__/selectors.test.ts:44]: The only way someone can reach this situation is if they manually fiddle with the space value in local storage, there's no valid workflow through which someone can reach this state.

And we could show an error but working case by case, in this particular case we are able to self-recover (by showing the base space), so IMO we should self-recover.

File: packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts (hunk 15 above)
Comment (develohpanda) [reply packages/insomnia-app/app/sync/git/__tests__/ne-db-client.test.ts:75]: It would matter - even though the `parentId` was not used for anything, the original value stored is `null`, and if we leave the `parentId` there depending on which space each individual user cloned into, git will always show changes. We need to resert this to what the previous value was.

See more context here: https://github.com/Kong/insomnia/blob/eb6795884660ef8a024af97af5d6c5857f0804c8/packages/insomnia-app/app/sync/ignore-keys.ts#L11-L17

File: packages/insomnia-app/app/plugins/context/data.ts (hunk 2 above)
Comment (develohpanda) [reply packages/insomnia-app/app/plugins/context/data.ts:29]: I do think there's a misunderstanding here and we don't need to move to a V5; this has nothing to do with the Insomnia export format but rather checking how frequently a fallback of "the plugin didn't provide a workspace so export all workspaces" is being used via this plugin API.

**Comment by develohpanda** [comment]
> However, I'd like to maybe start thinking about creating a new version of the insomnia format that accounts for spaces. That way we can ensure there will never be dangling workspaces (i.e. with null parentId like before). And that way we only need to check once upon booting the app (and then run a small migration if so) and then never again.

I'm honestly a bit confused about what impact a new Insomnia export format has on migrating workspaces on launch. The parentId is set during the process of importing a V4 file.

The two are unrelated items and I don't want to overlap them at all so curious where this is coming from; what is the issue (is it still a valid issue?), and why is an Insomnia V5 format the solution for it?

File: packages/insomnia-app/app/common/render.ts (hunk 20 above)
Comment (develohpanda) [reply packages/insomnia-app/app/common/render.ts]: [`0716ac3` (#3819)](https://github.com/Kong/insomnia/pull/3819/commits/0716ac34b125a4206ea6a09c89402ec2f1b85a8e)

File: packages/insomnia-app/app/plugins/context/data.ts (hunk 2 above)
Comment (develohpanda) [reply packages/insomnia-app/app/plugins/context/data.ts:29]: https://linear.app/insomnia/issue/INS-823/add-metrics-to-track-how-often-plugin-export-fallback-logic-is-being
